### ISSUE TITLE
Implement a bounding-box/volume-based approach for post processing

### ISSUE THREAD

**Issue by PrecisionRender** [godotengine/godot-proposals#4466]
### Describe the project you are working on

A 3D project that has multiple different themed areas.

### Describe the problem or limitation you are having in your project

The WorldEnvironment node and resource are global, to the entire scene. There is no way to set up a smaller area (e.g. inside a volcano) to have a different post processing setup to the rest of the scene.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Most game engines, like Unity and UE5, use a volume-based approach, where one can specify post processing settings inside of an AABB, that can modify a predefined area's post processing effects. Example:

> World has XY as global ambient light

> Smaller volume has WZ as local ambient light inside of predefined area

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Instead of a WorldEnvirinment node, there would be an EnvironmentVolume node, where one could specify the post processing effects that are active inside of the area. Or, one can turn off an option such as "Is Bounded" so that the EnvironmentVolume node acts as a WorldEnvironment node. Then one can put more EnvironmentVolume nodes into the Scene where one needs different post processing.

https://user-images.githubusercontent.com/89754713/165882903-bd9aa397-fbba-40da-9a88-083e3da4b3e8.mp4

Here is an example of a global post-process volume (left) and a local post-process volume (right).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Probably, yes, for a moderately experienced programmer. However, the people who use this tool to it's max potential are artists, who may not have the time or ability to create such a tool from scratch, especially when it's offered out of the box in other engines. 

Another plus to having such a node built-in is the ability for one to move the viewport camera into an EnvironmentVolume area and examine its effects in the editor, whereas if such a tool was scripted, post processing areas could only be examined at runtime, by changing WorldEnvironments dynamically through code.

And something that can't be easily achieved through GDScript or even GDExtension is having a proper fade between areas, so as not to have a sharp "pop" between different visual effects.

### Is there a reason why this should be core and not an add-on in the asset library?

See above.

**Comment by mrjustaguy** [comment]
This would be very useful for games that need transitions between land and water

**Comment by Calinou** [comment]
> Smaller volume has WZ as local ambient light inside of predefined area

This can be achieved with a ReflectionProbe node, which can smoothly override ambient light within a defined box (that can be rotated).

FogVolume nodes are also available for local fog, and they can use custom fog shaders as well.

**Comment by PrecisionRender** [comment]
> FogVolume nodes are also available in Godot 4.0 for local fog, and they can use custom fog shaders as well.

Yes, the FogVolume nodes are excellent. I was thinking of something similar for the entire WorldEnvironment.

**Comment by PrecisionRender** [comment]
> How would local skies work? I feel it doesn't make sense, at least from a background perspective.

I was just using it as an example. Skies don't necessarily need to change. 



> By "shader", I was referring to the GLSL shaders internally used in Godot to render various materials. It's important to keep those simple to improve performance and reduce shader compilation times.

That is important; something that Godot does well compared to other engines is shader compilation. I was suggesting doing the calculations in regular C++, then feeding that into a shader, like how the WorldEnvironment does. 

TL;DR: The EnvironmentVolume node decides what effects and parameters to use, then those are fed into the World3D. As far as the World3D is concerned, it's just one Environment resource, keeping all the shaders the same as they are now.

**Comment by clayjohn** [comment]
I imagine we could have something like a "LocalEnvironment" Node that overrides the current WorldEnvironment. We would have to limit it to only interpolate easily interpolatable properties (other properties like sky could just snap to the new resource when in the volume). I'm sure you could manage everything in the SceneTree well enough that you could avoid doubling any calculations in the rendering code.

**Comment by PrecisionRender** [comment]
> I imagine we could have something like a "LocalEnvironment" Node that overrides the current WorldEnvironment. We would have to limit it to only interpolate easily interpolatable properties (other properties like sky could just snap to the new resource when in the volume). I'm sure you could manage everything in the SceneTree well enough that you could avoid doubling any calculations in the rendering code.

This is also a good idea.

**Comment by Calinou** [comment]
Looking at the video[1], I finally understood what you mean. I was initially thinking that you'd expect different *pixels* to use different Environment properties, rather than affecting the entire view at once. This seems like a reasonable approach to me.

[1]: The video doesn't preview on Firefox (I get "No video with supported format and MIME type found."). I had to open it by right-clicking it and opening the copied video link in a new tab.

**Comment by PrecisionRender** [comment]
Ah, yes, I guess I was relying too much on the video to explain it for me. 😅 Sorry. Yes, it should effect the entire view based on the camera position.

**Comment by SIsilicon** [comment]
What if you give the already existing Area nodes an environment property?

**Comment by DarkKilauea** [comment]
I wrote an addon that addresses your requirements.  It acts much like the UE4 post processing volume, though it doesn't support blending between volumes yet.

Give it a try: https://github.com/DarkKilauea/godot-environment-volume

If there is interest in integrating it directly into Godot, I'd be willing to port it.

**Comment by PrecisionRender** [comment]
@DarkKilauea Your addon looks great! It appears like it adds pretty much everything an EnvironmentVolume node would need, minus blending.

If you can get blending between volumes working, this addon would be a great thing to add to the core.

**Comment by Calinou** [comment]
I like the idea of adding this functionality to Area, as it's already where audio bus overrides can be performed. This would also allow using arbitrary shapes rather than being limited to boxes (and perhaps ellipses).

**Comment by PrecisionRender** [comment]
> Time-based blending could be used as an alternative (e.g. crossfade over a time period defined in the Area node). It's much simpler to implement, but it doesn't work as well if the player can move slowly and backtrack around the Area.

@Calinou I think that time-based blending would work, as the game developer can adjust the fade-time based on the speed of the player if needed. I guess it's a question of whether custom shapes are worth more than distance-based blending.

**Comment by Zireael07** [comment]
+1 to having something available in the core for areas (I have a game using water areas, currently I fake things by having a roughly box shaped mesh and some shaders)

**Comment by PrecisionRender** [comment]
I'm good with Areas, too.

**Comment by mrjustaguy** [comment]
+1 for this to be in Areas.

How to solve blending - Make a Margin around the Collision shapes, have a way to determine how far you are from exiting/entering. all Primitive based Collision Shapes would work, and Convex would also be easy to calculate this, and Trimeshes aren't volumes anyway (at least not to the Engine)

**Comment by PrecisionRender** [comment]
@DarkKilauea I took another look at your plug-in. It looks great! Would you mind if I used it as a base to try to bring it into the engine?

Of course, if you'd rather do it yourself, then I won't.

**Comment by osayami** [comment]
When a similar request was made in [https://github.com/godotengine/godot/issues/21048](url) Reduz said 

>  This is one of the reasons why I wanted to split WorldEnvironment into effects and post processing eventually, so you can animate or switch between post processes and leave other parameters such as SSR/SSAO/etc alone.

here [https://github.com/godotengine/godot/issues/21048#issuecomment-425639647](url) . I would imagine this is not something that would be done in 4.x right?

**Comment by Calinou** [comment]
> I would imagine this is not something that would be done in 4.x right?

Not yet. This is being tracked in https://github.com/godotengine/godot-proposals/issues/2196, although I don't know if splitting built-in effects to be usable independently is still planned.

**Comment by Oblepikha** [comment]
Volume-based environments would be really nice. Not just a global override, but the ability
to render several different effects (fog, ambient) in different volumes.

There are a lot of cases where you need to create a dark interior within a bright outdoor environment.
Without some form of GI, ray tracing, or portal rendering, this is very hard to achieve.
Disabling sky contribution and ambient light breaks the outdoor part of the scene.
Reflection probes also introduce visible seams that are difficult to hidе. They also don't work well with lightmaps.

As an alternative, the scene can be split into multiple parts and lightmaps baked separately.

I managed to render the interior and exterior separately with different environments using rendering layers and a full-screen quad,
but this approach has many issues with transparent objects and depth precision.
<img width="1762" height="1192" alt="Image" src="https://github.com/user-attachments/assets/728e99d5-e498-4da1-8bbe-bb2c60e78a6e" />

For example, Flax Engine has a separate SkyLight entity with a radius for ambient lighting.

This is one of the cases where multipass rendering and stencil buffers could help.

**Comment by Raildex** [comment]
I would like to mention that my current workflow for "local environment" is like this:
Add Area3D to Camera3D with mask "EnvironmentArea".
Area3D reports to Camera via signal that an EnvironmentArea entered/exited.
The Camera3D has it's own `Environment` for the runtime.
This Environment gets tweened when an EnvironmentArea entered.

This has one serious drawback: No In-Editor preview of the post processing and I need to define have a fallback Environment when the last area exists.
I also need to manually specify *which* parts of the post processing are changed.

<img width="960" height="890" alt="Image" src="https://github.com/user-attachments/assets/d2559881-45ab-4937-a654-6f6968dfb50e" />

```gdscript
extends Area3D
class_name EnvironmentArea3D
@export var env: Environment
@export var fade_time: float = 0.0
@export var is_water: bool = false
enum AdjustmentFlags {
	BACKGROUND = 1 << 0,
	SKY = 1 << 1,
	AMBIENT = 1<< 2,
	REFLECTED_LIGHT = 1 << 3,
	TONEMAP = 1<<4,
	FOG = 1 << 5,
	VOLUMETRIC_FOG = 1 <<6,
	ADJUSTMENTS = 1<<7,
	GLOW = 1 << 8
}
@export_flags("Background","Sky","Ambient","Reflected Light","Tonemap","Fog","Volumetric Fog","Adjustments","Glow") var enabled_environment_changes: int

const PROPERTIES := {
	AdjustmentFlags.BACKGROUND : [
		^"background_energy_multiplier"
	],
	AdjustmentFlags.SKY : [
		^"sky",
		^"sky:sky_material",
		^"sky:sky_material:sky_top_color",
		^"sky:sky_material:sky_horizon_color",
		^"sky:sky_material:sky_curve",
		^"sky:sky_material:sky_energy_multiplier",
		^"sky:sky_material:sky_cover",
		^"sky:sky_material:sky_cover_modulate",
		^"sky:sky_material:ground_bottom_color",
		^"sky:sky_material:ground_horizon_color",
		^"sky:sky_material:ground_curve",
		^"sky:sky_material:ground_energy_multiplier",
		^"sky:sky_material:sun_angle_max",
		^"sky:sky_material:sun_curve",
		^"sky:sky_material:use_debanding",
		^"sky:sky_material:energy_multiplier",
	],
	AdjustmentFlags.AMBIENT : [
		^"ambient_light_sky_contribution",
		^"ambient_light_energy",
		^"ambient_light_color",
	],
	AdjustmentFlags.REFLECTED_LIGHT : [
		^"ambient_light_sky_contribution",
		^"ambient_light_energy",
		^"ambient_light_color",
	],
	AdjustmentFlags.TONEMAP : [
		^"tonemap_exposure",
		^"tonemap_white",
	],
	AdjustmentFlags.FOG : [
		^"fog_enabled",
		^"fog_mode",
		^"fog_light_color",
		^"fog_sun_scatter",
		^"fog_density",
		^"fog_aerial_perspective",
		^"fog_sky_affect",
		^"fog_height",
		^"fog_height_density",
		^"fog_depth_curve",
		^"fog_depth_begin",
		^"fog_depth_end",
	],
	AdjustmentFlags.VOLUMETRIC_FOG : [
		^"volumetric_fog_enabled",
		^"volumetric_fog_density",
		^"volumetric_fog_albedo",
		^"volumetric_fog_emission",
		^"volumetric_fog_emission_energy",
		^"volumetric_fog_gi_inject",
		^"volumetric_fog_anisotropy",
		^"volumetric_fog_length",
		^"volumetric_fog_detail_spread",
		^"volumetric_fog_ambient_inject",
		^"volumetric_fog_sky_affect",
	],
	AdjustmentFlags.ADJUSTMENTS : [
		^"adjustments_enabled",
		^"adjustment_brightness",
		^"adjustment_contrast",
		^"adjustment_saturation",
		^"adjustment_color_correction"
	],
	AdjustmentFlags.GLOW : [
		^"glow_enabled",
		^"glow_levels/1",
		^"glow_levels/2",
		^"glow_levels/3",
		^"glow_levels/4",
		^"glow_levels/5",
		^"glow_levels/6",
		^"glow_levels/7",
		^"glow_normalized",
		^"glow_intensity",
		^"glow_strength",
		^"glow_bloom",
		^"glow_blend_mode",
		^"glow_hdr_threshold",
		^"glow_hdr_scale",
		^"glow_hdr_luminance_cap",
		^"glow_map_strength"
	]
}

```


```gdscript

func _on_area_entered(area: Area3D) -> void:
	if area is EnvironmentArea3D:
		if not active_env_areas.has(area):
			active_env_areas.append(area)
		_sort_env_areas()
		_update_environment()


func _on_area_exited(area: Area3D) -> void:
	if area is EnvironmentArea3D:
		active_env_areas.erase(area)
		_update_environment()


func _sort_env_areas() -> void:
	active_env_areas.sort_custom(func(a: EnvironmentArea3D, b: EnvironmentArea3D):
		return a.priority < b.priority
	)


func _update_environment() -> void:
	var topmost_env: Environment
	var fade_time := 0.0
	var flags := 0xFFFFFFFF
	if active_env_areas.size() > 0:
		var env_area : EnvironmentArea3D = active_env_areas.back()
		topmost_env = env_area.env
		fade_time = env_area.fade_time
		flags = env_area.enabled_environment_changes
	if not topmost_env:
		return
	if flags == 0:
		return
	if env_tween:
		env_tween.kill()
	env_tween = create_tween()
	env_tween.set_parallel(true)
	for v in EnvironmentArea3D.AdjustmentFlags.values():
		if flags & v == v:
			for p: NodePath in EnvironmentArea3D.PROPERTIES[v]:
				var final_value = topmost_env.get(str(p))
				if final_value != null:
					env_tween.tween_property(
						cam.environment,
						p,
						final_value,
						fade_time
					)
			pass
	env_tween.play()
	#await env_tween.finished
	#cam.environment = topmost_env.duplicate_deep(Resource.DEEP_DUPLICATE_ALL)
	

```
